src/private_broadcast.cpp

```diff
@@ -14,10 +15,78 @@ void PrivateBroadcast::Add(const CTransactionRef& tx) EXCLUSIVE_LOCKS_REQUIRED(!
     }
 }
 
+std::optional<CTransactionRef> PrivateBroadcast::GetTxForBroadcast() EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+{
+    LOCK(m_mutex);
+    if (m_by_priority.empty()) {
+        return std::nullopt;
+    }
+    const Txid& txid = m_by_priority.begin()->second;
+    auto it = m_by_txid.find(txid);
+    if (Assume(it != m_by_txid.end())) {
+        return it->second.tx;
+    }
+    m_by_priority.erase(m_by_priority.begin());
+    return std::nullopt;
+}
+
+void PrivateBroadcast::PushedToNode(const NodeId& nodeid, const Txid& txid) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+{
+    LOCK(m_mutex);
+    m_by_nodeid.emplace(nodeid, txid);
+}
+
+bool PrivateBroadcast::BroadcastEnd(const NodeId& nodeid, bool confirmed_by_node) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+{
+    LOCK(m_mutex);
+    auto it = m_by_nodeid.find(nodeid);
+    if (it == m_by_nodeid.end()) {
+        return false;
+    }
+    const Txid txid{it->second};
+    m_by_nodeid.erase(it);
+
+    if (!confirmed_by_node) {
+        return true;
+    }
+
+    // Update broadcast stats.
+
+    auto iters = Find(txid);
+    if (!iters) {
+        return true;
+    }
+    Priority& priority = iters->by_txid->second.priority;
+
+    ++priority.num_broadcasted;
+    priority.last_broadcasted = GetTime<std::chrono::microseconds>();
+
+    m_by_priority.erase(iters->by_priority);
+    m_by_priority.emplace(priority, iters->by_txid->first);
```

Comment:
Yes, `iters->by_txid->first` is the same as `txid`, changed it to use `txid` which is shorter.

It is necessary to remove and re-add because we have changed the key and C++ maps do not support modifying of the keys. That is, changing the key means that the entry will have to be in another place in the map, so it has to be removed from the old place and added to the new one.